src/pool-bin/libraries/BinPool.sol

```diff
@@ -266,14 +269,120 @@ library BinPool {
         self.slot0.activeId = swapState.activeId;
 
         if (swapForY) {
-            uint128 consumed = amountIn - amountsLeft.decodeX();
+            uint128 consumed = amountIn - amountInLeft.decodeX();
             result = toBalanceDelta(-consumed.safeInt128(), amountsOut.decodeY().safeInt128());
         } else {
-            uint128 consumed = amountIn - amountsLeft.decodeY();
+            uint128 consumed = amountIn - amountInLeft.decodeY();
             result = toBalanceDelta(amountsOut.decodeX().safeInt128(), -(consumed.safeInt128()));
         }
     }
 
+    struct SwapParamsV2 {
+        bool swapForY;
+        uint16 binStep;
+        uint24 lpFeeOverride;
+        int128 amountSpecified; // negative for exactInput, positive for exactOutput
+    }
+
+    function swapV2(State storage self, SwapParamsV2 memory params)
+        internal
+        returns (BalanceDelta result, SwapState memory swapState)
+    {
+        Slot0 memory slot0Cache = self.slot0;
+        swapState.activeId = slot0Cache.activeId;
+        bool swapForY = params.swapForY;
+        swapState.protocolFee =
+            swapForY ? slot0Cache.protocolFee.getZeroForOneFee() : slot0Cache.protocolFee.getOneForZeroFee();
+        bool exactInput = params.amountSpecified < 0;
+
+        {
+            uint24 lpFee = params.lpFeeOverride.isOverride()
+                ? params.lpFeeOverride.removeOverrideAndValidate(LPFeeLibrary.TEN_PERCENT_FEE)
+                : slot0Cache.lpFee;
+
+            /// @dev swap fee includes protocolFee (charged first) and lpFee
+            swapState.swapFee = swapState.protocolFee == 0 ? lpFee : swapState.protocolFee.calculateSwapFee(lpFee);
+        }
+
+        /// @notice early return if hook has updated amountSpecified to 0
+        if (params.amountSpecified == 0) return (result, swapState);
+
+        uint128 amount = params.amountSpecified > 0 ? uint128(params.amountSpecified) : uint128(-params.amountSpecified);
+
+        /// @dev Amount of token left. In exactIn, refer to how much input left. In exactOut, refer to how much output left
+        bytes32 amountLeft = (swapForY == exactInput) ? amount.encodeFirst() : amount.encodeSecond();
```

Comment:
So this is basically unsigned version of amountSpecified, right ?

Reply:
also should we use amountsLeft i.e. (adding trailing s).  I guess that's the renaming convention for bin-pool

without s: it just a single value from one side either tokenX or tokenY
with s: it's a composition number which contains info about both sides

Reply:
yeh make sense updated